### DISCUSSION TITLE
First time contributing. Have a few questions.

### DISCUSSION THREAD

**Discussion by lcgeneralprojects** [py-pdf/fpdf2#1137] [Q&A] [answered]
Hello. I want to begin contributing to open-source projects. I have, however, hit a few obstacles here.

1. I am having trouble understanding how `OutputProducer.bufferize()` works in a short amount of time. How much should I know about it to begin solving issues and implementing features?
2. What tools do people use to make pdfs for testing? Is there anything else that I should know about the practice here that is not covered in `Development.md`?
3. By default, I am unable to import an editable library in other projects. How do other people deal with that?

**Comment by andersonhc** [comment, marked as answer]
Hi @lcgeneralprojects 

Thank you for reaching out and expressing your interest in contributing to `fpdf2`!

- Understanding OutputProducer.bufferize(): This function produces the final PDF file by processing and combining all the objects. However, for your first issue (#1073), you won't need to interact with output.py.
- Tools for Making PDFs: I typically use don't have to use anything other than `fpdf2`. You can follow the minimal code provided in the issue description to generate a test PDF file. Alternatively, when creating your own tests, you can use assert_pdf_equal() with generate=True to produce a PDF file.
- Dealing with Editable Libraries: I normally don't have problem with editable libraries except when using VS Code. If you're encountering problems specifically with VS Code, there are a couple of solutions: adding the library location to the "extraPaths" or installing the library in editable mode with the "strict" parameter. For a more detailed explanation, this Stack Overflow discussion might be helpful: [Python packages imported in editable mode can't be resolved by Pylance in VSCode](https://stackoverflow.com/questions/76213501/python-packages-imported-in-editable-mode-cant-be-resolved-by-pylance-in-vscode).

Please, don't hesitate to ask if you have any more questions or need further clarification on anything.

**Reply by lcgeneralprojects** [reply]
Thank you for a quick response. Alas, I was not able to read it in full yesterday.

> Alternatively, when creating your own tests, you can use assert_pdf_equal() with generate=True to produce a PDF file

Hmm. I will need to experiment with it, it seems, but I currently do not understand how it would generate a needed file for `assert` comparisons. I assumed that one would need to manually create them (including in the form of SVGs that are to be converted into PDF file), given that when one wants to test a feature that is to be implemented, a creation of such a file using only `fpdf2` seems to not necessarily be possible.

> Dealing with Editable Libraries: I normally don't have problem with editable libraries except when using VS Code

I have encountered the issue in VS Code and Pycharm. I currently deal with it by attaching the directory with the library to the relevant projects as sources. I will go over the Stack Overflow discussion.

> Please, don't hesitate to ask if you have any more questions or need further clarification on anything

Thank you, I will.
